### ISSUE TITLE
include_role and meta dependencies: strange behaviour on conditionals

### ISSUE THREAD

**Issue by philfry** [ansible/ansible#23733]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
role dependencies

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan 12 2017, 17:59:37) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```

##### CONFIGURATION
n/a

##### OS / ENVIRONMENT
n/a

##### SUMMARY
an `include_role` gets skipped when a former `dependency` condition is not met

##### STEPS TO REPRODUCE
```bash
echo -e "[defaults]\ninventory = ./inventory\nnocows = 1" > ansible.cfg
echo "localhost ansible_connection=local" > inventory
mkdir -p roles/{dep{1..3},master}/{files,templates,defaults,meta,handlers,vars,tasks}
echo -e "---\n- debug: msg='in {{role_name}}'" | tee roles/dep{1..3}/tasks/main.yml > /dev/null
echo -e "---\ndependencies:\n  - role: dep1\n  - role: dep2\n    when: false" > roles/master/meta/main.yml              
echo -e "---\n- include_role: name=dep3"  > roles/master/tasks/main.yml
echo -e "---\n- hosts: [ 'localhost' ]\n  gather_facts: no\n  roles: [ 'master' ]" > bug.yml
ansible-playbook bug.yml
```

##### EXPECTED RESULTS
```json
ok: [localhost] => {
    "changed": false, 
    "msg": "in dep1"
}
ok: [localhost] => {
    "changed": false, 
    "msg": "in dep3"
}
```

##### ACTUAL RESULTS
```json
ok: [localhost] => {
    "changed": false, 
    "msg": "in dep1"
}
```
verbose output in gist: https://gist.github.com/philfry/8099cc98e12d7b36bf015dd856b2c3cb

**Comment by snowsky** [comment]
I like the command lines which is easy to reproduce the issue.

**Comment by snowsky** [comment]
no idea why dep3 is lost

**Comment by snowsky** [comment]
The issue is caused by when cause.

**Comment by eerorika** [comment]
It seems to me that this could be the same bug I reported earlier on 2.2: https://github.com/ansible/ansible/issues/23379

**Comment by alikins** [comment]
(repro expanded for reference)

ansible.cfg
``` ini
[defaults]\ninventory = ./inventory
nocows = 1
```

bug.yml
``` yaml
---
- hosts: [ 'localhost' ]
  gather_facts: no
  roles: [ 'master' ]
```

roles/master/meta/main.yml
``` yaml
---
dependencies:
  - role: dep1
  - role: dep2
    when: false
```

roles/master/tasks/main.yml
``` yaml
---
- include_role: name=dep3
```

roles/dep*/tasks/main.yml
``` yaml
---
- debug: msg='in {{role_name}}'
```

**Comment by philfry** [comment]
The bug only occurs when using `include_role` in a role.
This example works fine:

- roles/master/tasks/main.yml
```yaml
---
- debug: msg="in master"
```
- bug.yml
```yaml
---
- hosts: [ 'localhost' ]
  gather_facts: no
  roles: [ 'master' ]
  tasks:
    - include_role: name=dep3
```
### results
```json
TASK [dep1 : task dep1]
ok: [localhost] => {
    "msg": "in dep1"
}

TASK [dep2 : task dep2]
skipping: [localhost]

TASK [master : task master]
ok: [localhost] => {
    "msg": "in master"
}

TASK [dep3 : task dep3]
ok: [localhost] => {
    "msg": "in dep3"
}
```
While playing around with ansible's source code and adding some debug lines I found the following:
```
playbook/role/__init__.py:_load_dependencies
  -> playbook/role/__init__.py:load
    -> playbook/role/__init__.py:_load_role_data
       -> playbook/helpers.py:load_list_of_blocks
          found TASK: dep1 : debug, when is []
    -> playbook/role/__init__.py:_load_role_data
       -> playbook/helpers.py:load_list_of_blocks
          found TASK: dep2 : debug, when is [False]
playbook/helpers.py:load_list_of_tasks
  -> playbook/role_include.py:get_block_list
     found TASK: master : include_role
playbook/role/__init__.py:_load_role_data
  -> playbook/helpers.py:load_list_of_blocks
     found task TASK: dep3 : debug, when is []
playbook/role_include.py:get_block_list
  -> playbook/role/__init__.py:compile
     task block [TASK: dep3 : debug], when is []
# so it's still correct here.

playbook/role/__init__.py:_load_role_data
  -> playbook/helpers.py:load_list_of_blocks
     found block BLOCK(uuid=3c18a002-ba38-edb0-5533-000000000015)(id=139940181569616)(parent=TASK: master : include_role), when is [False]
        found task TASK: dep3 : debug, when is [False]
# here it is, probably.
```
full debug log is [here](https://gist.github.com/ac1c8021a035ac3f7aa412976baeb232) with [these modifications](https://gist.github.com/1ad81f4d9b6560f0d52405edaf9c98c7)

**Comment by sivel** [comment]
This issue was resolved in eeaa67657750005b5ff9c287b1883fc08692814d and included in the v2.3.2 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
